src/collection-view.js

```diff
@@ -125,11 +125,75 @@ const CollectionView = Backbone.View.extend({
     }
   },
 
-  // get the child view by model it holds, and remove it
-  _onCollectionRemove(model) {
-    const view = this.children.findByModel(model);
-    this.removeChildView(view);
-    this._checkEmpty();
+  // Handle collection update model removals
+  _onCollectionUpdate(collection, {changes: {removed = []} = {}} = {}) {
+    if (_.isArray(removed) && removed.length) {
+      this._removeChildViews(removed);
+    }
+  },
+
+  // Remove the child views and destroy them.
+  // This function also updates the indices of later views
+  // in the collection in order to keep the children in sync with the collection.
+  // Return array of views where the first view has the index that
+  // tells CollectionView#_updateIndices where to start syncing remaining
+  // view indices.
+  _removeChildViews(modelsOrViews, {checkEmpty} = {}) {
+    if (!modelsOrViews.length) {
+      return modelsOrViews;
+    }
+
+    const shouldCheckEmpty = checkEmpty !== false;
+
+    // Used to determine where to update the remaining
+    // sibling view indices after these views are removed.
+    const removedViews = _.reduce(modelsOrViews, (removingViews, modelOrView) => {
```

Comment:
This is a large iterator.  It'd be nice to move it to it's own function so that this line is just `const removeViews = _.reduce(modelsOrViews, (removingViews, modelOrView) => this._removeChildView, []);`